[August] 미션3: 체스판 초기화

polynomeer opened the pull request:
### 미션3: 체스판 초기화

안녕하세요! '미션3: 체스판 초기화'를 구현하고 주어진 테스트 코드를 실행 해보았습니다. 구현 내용은 아래와 같습니다.

### 구현 내용

#### main/pieces

- Piece.java
  - char 타입의 representation필드를 추가하여, 상속받은 체스말들이 모두 표기될 문자를 필드로 갖도록 했습니다.
  - representation에 대한 getter도 추가했습니다.
   
- Pawn.java
  - 상수 필드로 검정색 폰, 흰색 폰의 표기문자를 갖습니다.
  - 생성자 오버로딩을 통해 기본적으로는 흰색의 값을 가지도록 했습니다.
  - getRepresentation()은 부모 생성자(Piece의 생성자)를 호출해서 representation필드를 읽어옵니다.

#### main/chess

- Board.java

  - 기존의 board 리스트를 `whitePieces`, `blackPieces`로 각각 나누어서 저장하도록 했습니다.
  - 기존의 add 메서드를 색깔별로 `addWhite`, `addBlack`으로 나누었습니다.
  - find메서드도 `findWhite`, `findBlack`으로 나누었습니다.
  - `initialize()` : `BOARD_SIZE`라는 상수를 필드에 선언하여, 그 크기만큼 체스말들을 리스트에 추가합니다.
  - `getPiecesResult(List<Piece) pieces)` : 체스말 리스트를 매개변수로 받아서 해당 말의 표기문자를 문자열에 하나씩 붙여서 리턴합니다.
  - `getBlackPawnsResult()`와 `getWhitePawnsResult()`는 `getPiecesResult()`를 호출하는데, 색깔별 체스말의 리스트를 넘겨줍니다. 추후에는 `getBlackResult()`, `getWhiteResult()`로 변경하여 전체 결과를 가져올 수 있도록 해야합니다.
  - `print()` : 폰 이외에 빈칸을 `.`으로 모두 채우고 폰의 위치에 `getWhitePawnsResult()`, `getBlackPawnsResult()`로 가져온 결과 문자열을 이어붙여서 전체 보드를 출력합니다.

  

#### test/chess

- BoardTest.java
  - `initialize()` : `Board::initialize()`호출 후 각 색깔의 폰 리스트가 적절히 초기화 되었는지 테스트합니다.
  - 이전에 작성한 메서드들은 우선은 if문으로 분기하여 실행할 수 있도록 변경했습니다.
  - `assertEquals()` 보다 `assertThat()`이 장점이 많다는 것을 학습하여 `assertThat()`으로 변경했습니다.

polynomeer committed:
feat: Add representation field in Piece class

Added char representation field in Piece class.
Added overloaded constructors in Pawn class.
Changed test code in PawnTest.

polynomeer committed:
feat: Add getter for representation field

Added getter for representation field in Piece class.
getRepresentation() in Pawn class call Piece's(super's) getter.
Added test code in create_default() to test default constructor.

polynomeer committed:
feat: Add test code in BoardTest

Added initialize() in BoardTest.
Defined initialize(), getWhitePawnsResult(), getBlackPawnsResult() in
Board class.

polynomeer committed:
feat: Add black and white methods in Board

Added addBlack() and addWhite() to add piece by color.
Added findBlack() and findWhite() to find by color.
Implemented initialize() to initialize board.

polynomeer committed:
feat: Separated add into addBlack and addWhite

Separated add() into addBlack() and addWhite().
Changed test code in BoardTest class and passed it.

polynomeer committed:
chore: Change assertEquals to assertThat

polynomeer committed:
feat: Add Game class (Main class)

Added Game class and it include main game logic.
While loop is iterate until command is equal to "end".
If command is equal to "start", construct Board and initialize the
board.

tmdgusya commented on src/main/java/chess/Game.java:
```diff
@@ -0,0 +1,30 @@
+            command = command.toLowerCase();
+            if (command.equals("start")) {
+                Board board = new Board();
+                board.initialize();
```
지금 이렇게 되면 initialize() 가 계속해서 반복되지 않나요?
만약 저사이에서 play() 를 해야되는 상황이면 말을 움직이게 된다해도, 초기화 될것 같이 보입니다.

while 문 밖으로 빼서 초기에 한번만 실행되게 해도 좋을 것 같습니다.

sonypark commented on src/main/java/chess/Game.java:
```diff
@@ -0,0 +1,30 @@
+            }
+        }
+
+        System.out.println("Goodbye..");
+        scanner.close();
```
👍

sonypark commented on src/main/java/chess/Board.java:
```diff
@@ -1,22 +1,69 @@
 
-    public void add(Piece piece) {
-        board.add(piece);
+    public void addBlack(Piece piece) {
+        blackPieces.add(piece);
+    }
```
color가 white인 Piece가 인자로 들어오면 어떻게 되나요?

polynomeer replied:
color에 대한 검증이 없네요! 혹시 add를 하나로 합쳐서 분기하는 것이 더 나은 방법 일까요? 계속 고민했는데 모르겠네요.

polynomeer replied:
add와 find 메서드는 색깔별로 따로 있는게 맞다고 생각해서 검증하는 코드만 추가했습니다! `BoardTest`에 있는 검증 코드도 개별 테스트 및 assertAll()을 사용할 수 있도록 각각 분리하였습니다.

sonypark commented on src/test/java/chess/BoardTest.java:
```diff
@@ -2,43 +2,65 @@
+    @Test
+    @DisplayName("Init board")
+    public void initialize() throws Exception {
+        Board board = new Board();
+        board.initialize();
```
BeforeEach에서 생성한 board를 사용하는건 어떨까요?

polynomeer replied:
앗, 중복이네요! 테스트 코드에 좀 더 신경을 쓰도록 하겠습니다.

sonypark commented on src/test/java/pieces/PawnTest.java:
```diff
@@ -4,29 +4,30 @@
 
-    void verifyPawn(final String color) {
-        Pawn pawn = new Pawn(color);
+    void verifyPawn(final String color, final char representation) {
+        Pawn pawn = new Pawn(color, representation);
         assertThat(pawn.getColor()).isEqualTo(color);
     }
```
color에 대한 검증만 하고 있네요

polynomeer replied:
처음에 무슨말인지 몰랐는데, representation에 대한 검증을 하는 메서드를 따로 구현하여 각각 검증하도록 했습니다!

sonypark commented on src/main/java/pieces/Pawn.java:
```diff
@@ -1,16 +1,26 @@
 package pieces;
 
 public class Pawn extends Piece {
+    public static final char BLACK_REPRESENTATION = 'P';
+    public static final char WHITE_REPRESENTATION = 'p';
```
👍

sonypark commented on src/main/java/chess/Board.java:
```diff
@@ -1,22 +1,69 @@
+        }
+        return sb.toString();
+    }
+
+    void print() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("........\n");
+        sb.append(getBlackPawnsResult()).append("\n");
+        for (int i = 0; i < BOARD_SIZE - 4; i++) {
+            sb.append("........\n");
+        }
+        sb.append(getWhitePawnsResult()).append("\n");
+        sb.append("........\n");
+
+        System.out.println(sb.toString());
```
출력 로직이 Board가 수행할 역할이 맞을지도 고민해보시면 좋을 것 같아요~

polynomeer replied:
클래스의 역할과 맞지 않는 것 같아서, `Game`에서 담당하도록 변경해 보았습니다!

sonypark replied:
네 입출력 로직을 담당하는 클래스를 만드는 방법도 있습니다. 진행하면서 필요할 때 분리하는것도 고려해보시면 좋겠네요!

polynomeer replied:
입출력 로직이 많아진다면 Printer등의 클래스를 별도로 구현해보겠습니다. 감사합니다! 😀

sonypark reviewed (changes requested):
안녕하세요 August, 리뷰어 소니 입니다.😊
요구사항에 맞게 잘 구현해주셨네요!
커밋 메시지를 잘 작성해주셔서 좋았습니다 👍
코멘트 확인하시고 객체의 역할, 코드 재사용성 부분을 고민해보시면 좋을 것 같습니다.

polynomeer committed:
refactor: Remove local variable 'board' in BoardTest

Removed local variable 'board' in initialize() block.
static variable which initialized in `@BeforeEach` substitutes it.

polynomeer committed:
feat: Add verifyPawnRepresentation method in PawnTest class

Added verifyPawnRepresentation() in PawnTest class to verify each
representation of pawn.
Added verifyPawnColor() verifies pawn color using overloaded
constructor.
I did not add overloaded constructor for pawn representation because
maybe the case that construct without color will not be occurred.

polynomeer committed:
refactor: Move print method from Board to Game

Moved print() from Board class to Game class to be consistent with class
role.
Added process() in Game class for separation of responsibilities.

polynomeer committed:
feat: Add color exception handling code in Board

Added logic verifying color in addWhite() and addBlack() methods in Board class to handle color exception.
Separated all verifying methods to use assertAll() in BoardTest class.

sonypark reviewed (approved):
요구사항 반영 확인했습니다.
다음 단계 진행해주세요~💪
